### PR TITLE
spinner, dialog box, response component, interceptor

### PR DESCRIPTION
**Pull request by syedameenabrar**

### CODE REVIEW

File: kb-survey/angular.json [hunk 1]
```diff
@@ -102,12 +103,16 @@
               "src/assets"
             ],
             "styles": [
+              "./node_modules/@angular/material/prebuilt-themes/indigo-pink.css",
```
Comment (5Amogh) [inline kb-survey/angular.json]: why are we adding styles multiple times here?

File: kb-survey/angular.json [hunk 2]
```diff
@@ -102,12 +103,16 @@
...
               "src/styles.scss"
             ],
             "scripts": []
           }
         }
       }
     }
+  },
+  "cli": {
```
Comment (5Amogh) [inline kb-survey/angular.json]: Please remove this analytics

File: kb-survey/package.json [hunk 3]
```diff
@@ -11,6 +11,7 @@
   "private": true,
   "dependencies": {
     "@angular/animations": "^14.2.0",
+    "@angular/cdk": "^14.2.0",
```
Comment (5Amogh) [inline kb-survey/package.json]: why are we installing angular cdk?

File: kb-survey/src/app/services/base-api/api-base.service.ts [hunk 4]
```diff
@@ -1,7 +1,7 @@
 import { Injectable } from '@angular/core';
 import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
 import { Observable } from 'rxjs';
-import { environment } from 'src/environments/environment';
+import { environment } from 'src/environments/environment.sample';
```
Comment (5Amogh) [inline kb-survey/src/app/services/base-api/api-base.service.ts]: Revert this change please

File: kb-survey/src/app/services/interceptor/api-interceptor.service.ts [hunk 5]
```diff
@@ -1,13 +1,51 @@
...
 @Injectable({
   providedIn: 'root'
 })
 export class ApiInterceptorService implements HttpInterceptor {
+  constructor(private responseService: ResponseService, private router:Router,
+    public dialog: MatDialog){
+
+  }
+
+  openConfirmationDialog(title: any, message: any, timer: any, actionBtns: boolean,
```
Comment (5Amogh) [inline kb-survey/src/app/services/interceptor/api-interceptor.service.ts]: in this method instead of sending arguments one by one, you can send one object with all these properties

File: kb-survey/src/app/services/interceptor/api-interceptor.service.ts [hunk 6]
```diff
@@ -18,6 +56,25 @@ export class ApiInterceptorService implements HttpInterceptor {
     });
 
     return next.handle(modifiedRequest).pipe(
+      tap(event => {
```
Comment (5Amogh) [inline kb-survey/src/app/services/interceptor/api-interceptor.service.ts]: why is this code commented?

File: kb-survey/src/app/shared/components/dialog/dialog.component.scss [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+.success {
+  color: green;
+}
+
+.error {
+  color: red;
+}
+
+.borderRadius {
+  border: 1px solid #6a1b9a !important;
```
Comment (5Amogh) [inline kb-survey/src/app/shared/components/dialog/dialog.component.scss]: the border radius has hardcoded color, please make it dynamic by using global color variables.. !important is it necessary here? if not pls remove

File: kb-survey/src/app/shared/components/dialog/dialog.component.scss [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+/* component-specific styles */
+
+/* Set the dialog box to min-width: max-content */
+.mat-dialog-container {
+  min-width: max-content;
+}
+
+/* Customize button styles */
+.mat-dialog-actions .purple-button {
+  background-color: #6a1b9a;
```
Comment (5Amogh) [inline kb-survey/src/app/shared/components/dialog/dialog.component.scss]: use var(--primary-color) instead of hardcoding the color

File: kb-survey/src/app/shared/components/dialog/dialog.component.scss [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+  /* Purple */
+  color: #fff;
```
Comment (5Amogh) [inline kb-survey/src/app/shared/components/dialog/dialog.component.scss]: same feedback as above

File: kb-survey/src/app/shared/components/dialog/dialog.component.scss [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+.mat-dialog-actions .purple-border-button {
+  background-color: transparent;
+  border: 2px solid #6a1b9a;
+  /* Purple */
+  color: #6a1b9a;
```
Comment (5Amogh) [inline kb-survey/src/app/shared/components/dialog/dialog.component.scss]: same feedback as above

File: kb-survey/src/app/shared/components/dialog/dialog.component.scss [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+  /* Purple */
+}
+
+.mat-dialog-actions .purple-border-button:hover {
+  background-color: rgba(106, 27, 154, 0.1);
```
Comment (5Amogh) [inline kb-survey/src/app/shared/components/dialog/dialog.component.scss]: pls check if you can use global color variable here as well

File: kb-survey/src/app/shared/components/dialog/dialog.component.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+  templateUrl: './dialog.component.html',
+  styleUrls: ['./dialog.component.scss']
+})
+export class DialogComponent {
+  constructor(
+    public dialogRef: MatDialogRef<DialogComponent>,
+    @Inject(MAT_DIALOG_DATA) public data: any
+  ) { }
+
+  onNoClick(): void {
```
Comment (5Amogh) [inline kb-survey/src/app/shared/components/dialog/dialog.component.ts]: what is this onNoClick()? and also pls rename the variable which would be easy to understand what its doing

File: kb-survey/src/app/shared/components/response/response.component.ts [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+@Component({
+  selector: 'app-response',
+  templateUrl: './response.component.html',
+  styleUrls: ['./response.component.scss']
+})
+export class ResponseComponent implements OnInit {
+  response:any = "Page Not Found"
+  constructor(private responseService:ResponseService) { 
+    this.responseService.sendResponse.subscribe((value) => {
```
Comment (5Amogh) [inline kb-survey/src/app/shared/components/response/response.component.ts]: Why are we subscribing inside the constructor instead of ngOInit any particular reason? any advantage of doing that? and if there's no use of ngOnInit you can remove that method

File: kb-survey/src/app/shared/components/spinner/spinner.component.html [hunk 14]
```diff
@@ -0,0 +1,5 @@
+<div class="spinner-overlay">
+    <mat-progress-spinner class="example-margin spinner" [color]="color" [mode]="mode" [value]="value"
```
Comment (5Amogh) [inline kb-survey/src/app/shared/components/spinner/spinner.component.html]: what is example margin?

File: kb-survey/src/app/shared/components/spinner/spinner.component.ts [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+
+@Component({
+  selector: 'app-spinner',
+  templateUrl: './spinner.component.html',
+  styleUrls: ['./spinner.component.scss']
+})
+export class SpinnerComponent implements OnInit {
+  color: ThemePalette = 'primary';
+  mode: ProgressSpinnerMode = 'indeterminate';
+  value = 50;
```
Comment (5Amogh) [inline kb-survey/src/app/shared/components/spinner/spinner.component.ts]: what does value indicate here? can you pls explain

File: kb-survey/src/app/shared/components/spinner/spinner.component.ts [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+  strokeWidth = 4;
+  constructor() { }
+
+  ngOnInit(): void {
```
Comment (5Amogh) [inline kb-survey/src/app/shared/components/spinner/spinner.component.ts]: If we are not using ngOnInit lifecycle then pls remove it

File: kb-survey/src/app/survey/survey.component.html [hunk 17]
```diff
@@ -4,4 +4,4 @@
   [assessment]="assessmentResult"
   (submitOrSaveEvent)="submitOrSaveEvent($event)"
   [fileuploadresponse]="fileUploadResponse"
-></questionnaire-player-main>
\ No newline at end of file
+></questionnaire-player-main>
```
Comment (5Amogh) [inline kb-survey/src/app/survey/survey.component.html]: pls revert this change, looks like spacing change

File: kb-survey/src/app/survey/survey.component.ts [hunk 18]
```diff
@@ -41,8 +49,40 @@ export class SurveyComponent implements OnInit {
       this.profileDetails = detailsOfProfile;
     });
 
+    this.route.queryParams.subscribe((queryParam) => {
```
Comment (5Amogh) [inline kb-survey/src/app/survey/survey.component.ts]: Pls keep this and remove the above code from line number 47 to 50

File: kb-survey/src/app/survey/survey.component.ts [hunk 19]
```diff
@@ -41,8 +49,40 @@ export class SurveyComponent implements OnInit {
...
+      this.profileDetails = queryParam;
+    });
     this.fetchSurveyDetails();
   }
+  openConfirmationDialog(title: any, message: any, timer: any, actionBtns: boolean,
```
Comment (5Amogh) [inline kb-survey/src/app/survey/survey.component.ts]: same feedback as i have given somewhere above for this function

File: kb-survey/src/app/survey/survey.component.ts [hunk 20]
```diff
@@ -101,30 +141,82 @@ export class SurveyComponent implements OnInit {
...
         console.log(this.assessmentResult);
       });
   }
 
-  submitOrSaveEvent(event: any): void {
-    const evidenceData = { ...event.detail.data, status: event.detail.status };
-    this.baseApiService
-      .post(
-        urlConfig.surveySubmissionURL +
+  async submitOrSaveEvent(event: any) {
```
Comment (5Amogh) [inline kb-survey/src/app/survey/survey.component.ts:150]: Please modify the logic here, try making it generic instead of calling the whole logic just to differentiate save and submit

File: kb-survey/src/environments/environment.sample.ts [hunk 21]
```diff
@@ -1,5 +1,5 @@
 export const environment = {
   production: false,
-  baseURL:'Insert your base url'
+  baseURL:"https://9593-106-51-188-161.ngrok-free.app"
```
Comment (5Amogh) [inline kb-survey/src/environments/environment.sample.ts]: pls delete this file, environment sample file is kept so you can copy the same structure and run the application using different env url in local

File: kb-survey/src/index.html [hunk 22]
```diff
@@ -7,8 +7,11 @@
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <link rel="icon" type="image/x-icon" href="favicon.ico">
   <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
+  <link rel="preconnect" href="https://fonts.gstatic.com">
+  <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@300;400;500&display=swap" rel="stylesheet">
+  <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.2/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-T3c6CoIi6uLrA9TneNEoa7RxnatzjcDSCmG1MXxSR1GAsXEV/Dwwykc2MPK8M2HN" crossorigin="anonymous">
```
Comment (5Amogh) [inline kb-survey/src/index.html]: why are adding bootstrap here? pls remove

File: kb-survey/src/index.html [hunk 23]
```diff
@@ -7,8 +7,11 @@
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <link rel="icon" type="image/x-icon" href="favicon.ico">
   <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
+  <link rel="preconnect" href="https://fonts.gstatic.com">
```
Comment (5Amogh) [inline kb-survey/src/index.html]: pls check and remove this gstatic link

File: kb-survey/src/index.html [hunk 24]
```diff
@@ -7,8 +7,11 @@
...
+  <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@300;400;500&display=swap" rel="stylesheet">
+  <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.2/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-T3c6CoIi6uLrA9TneNEoa7RxnatzjcDSCmG1MXxSR1GAsXEV/Dwwykc2MPK8M2HN" crossorigin="anonymous">
 </head>
-<body>
+<body class="mat-typography">
```
Comment (5Amogh) [inline kb-survey/src/index.html]: why are we doing this here?

File: kb-survey/src/main.ts [hunk 25]
```diff
@@ -2,7 +2,7 @@ import { enableProdMode } from '@angular/core';
 import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 
 import { AppModule } from './app/app.module';
-import { environment } from './environments/environment';
+import { environment } from './environments/environment.sample';
```
Comment (5Amogh) [inline kb-survey/src/main.ts]: pls revert this change

File: kb-survey/src/styles.scss [hunk 26]
```diff
@@ -19,3 +19,6 @@
     stroke: var(--primary-color);
 }
 
+
+html, body { height: 100%; }
```
Comment (5Amogh) [inline kb-survey/src/styles.scss]: why do need these changes?

File: README.md
Comment (5Amogh) [inline README.md:1]: why did we remove this file?

File: kb-survey/src/app/app.component.html [hunk 27]
```diff
@@ -1 +1 @@
-<router-outlet></router-outlet>
\ No newline at end of file
+<router-outlet></router-outlet>
```
Comment (5Amogh) [inline kb-survey/src/app/app.component.html]: looks like spacing issue, pls revert the change

**Review by 5Amogh — CHANGES_REQUESTED** [review]

File: kb-survey/angular.json (hunk 1 above)
Comment (syedameenabrar) [reply kb-survey/angular.json]: i have not imported it, i got conflicts when i took pull, i thought u have made some changes

File: kb-survey/package.json (hunk 3 above)
Comment (syedameenabrar) [reply kb-survey/package.json]: this also added while taking pull request

File: kb-survey/angular.json (hunk 1 above)
Comment (5Amogh) [reply kb-survey/angular.json]: check and remove if there's duplication and see if everything works fine..

File: kb-survey/src/app/services/base-api/api-base.service.ts (hunk 4 above)
Comment (syedameenabrar) [reply kb-survey/src/app/services/base-api/api-base.service.ts]: For this there is no file with name environment.ts

File: kb-survey/src/app/services/interceptor/api-interceptor.service.ts (hunk 5 above)
Comment (syedameenabrar) [reply kb-survey/src/app/services/interceptor/api-interceptor.service.ts]: ok i will work on it

File: kb-survey/src/index.html (hunk 24 above)
Comment (syedameenabrar) [reply kb-survey/src/index.html]: i have not added it

File: kb-survey/src/main.ts (hunk 25 above)
Comment (syedameenabrar) [reply kb-survey/src/main.ts]: there is no file with name evironment.ts

File: kb-survey/package.json (hunk 3 above)
Comment (5Amogh) [reply kb-survey/package.json]: sure? cause its not there in the repo, that's why its showing as new change.. either way remove it

File: kb-survey/src/app/services/base-api/api-base.service.ts (hunk 4 above)
Comment (5Amogh) [reply kb-survey/src/app/services/base-api/api-base.service.ts]: yes that's what we need to create using environemnt.sample.ts @syedameenabrar  in our local only.

File: kb-survey/src/styles.scss (hunk 26 above)
Comment (syedameenabrar) [reply kb-survey/src/styles.scss]: it was added for spinner color, i will check this if not nacessary i will remove it

File: README.md
Comment (syedameenabrar) [reply README.md:1]: i did not remove it

File: kb-survey/src/app/app.component.html (hunk 27 above)
Comment (syedameenabrar) [reply kb-survey/src/app/app.component.html]: ok

File: kb-survey/src/main.ts (hunk 25 above)
Comment (5Amogh) [reply kb-survey/src/main.ts]: yes that's what we need to create using environemnt.sample.ts @syedameenabrar in our local only.

File: README.md
Comment (5Amogh) [reply README.md:1]: might have removed accidentally pls add it back

File: kb-survey/src/styles.scss (hunk 26 above)
Comment (5Amogh) [reply kb-survey/src/styles.scss]: okay

File: kb-survey/angular.json (hunk 1 above)
Comment (syedameenabrar) [reply kb-survey/angular.json]: i have removed it

File: kb-survey/angular.json (hunk 2 above)
Comment (syedameenabrar) [reply kb-survey/angular.json]: removed it

File: kb-survey/package.json (hunk 3 above)
Comment (syedameenabrar) [reply kb-survey/package.json]: removed it

File: kb-survey/src/app/services/interceptor/api-interceptor.service.ts (hunk 6 above)
Comment (syedameenabrar) [reply kb-survey/src/app/services/interceptor/api-interceptor.service.ts]: i have removed commented code

File: kb-survey/src/styles.scss (hunk 26 above)
Comment (syedameenabrar) [reply kb-survey/src/styles.scss]: i have removed it

File: README.md
Comment (5Amogh) [inline README.md:1]: Why is it still showing this file as deleted can you pls check?

File: kb-survey/package-lock.json
Comment (5Amogh) [inline kb-survey/package-lock.json:1]: Can you please revert these changes as well?

File: kb-survey/src/app/shared/components/dialog/dialog.component.html [hunk 28]
```diff
@@ -0,0 +1,11 @@
+<h2 mat-dialog-title [ngClass]="{'success':data?.title == 'Success', 'error':data?.title == 'Error'}">{{data?.title}}</h2>
+<mat-dialog-content>
+    {{data?.message}}
+</mat-dialog-content>
+<div  *ngIf="data?.actionBtns">
+  <hr class="partition">
+<mat-dialog-actions>
+  <button mat-button (click)="onCancle()" class="purple-border-button borderRadius" mat-dialog-close>{{data?.btnLeftLabel}}</button>
+  <button mat-button [mat-dialog-close]="true" class="purple-button borderRadius" cdkFocusInitial>{{data?.btnRightLabel}}</button>
```
Comment (5Amogh) [inline kb-survey/src/app/shared/components/dialog/dialog.component.html]: cdkFocusInitial, why are we using this?

File: kb-survey/src/app/survey/survey.component.ts [hunk 29]
```diff
@@ -37,12 +45,38 @@ export class SurveyComponent implements OnInit {
...
+        title: confirmationParams?.title,
+        message: confirmationParams?.message,
+        actionBtns: confirmationParams?.actionBtns,
+        btnLeftLabel: confirmationParams?.btnLeftLabel,
+        btnRightLabel: confirmationParams?.btnRightLabel,
+      }
+    });
+
+    dialogRef.afterClosed().subscribe(result => {
+      if (result) {
```
Comment (5Amogh) [inline kb-survey/src/app/survey/survey.component.ts]: instead of usin if-esle here, you can use ternary operator for this, code would be much cleaner

File: kb-survey/src/app/survey/survey.component.ts [hunk 30]
```diff
@@ -37,12 +45,38 @@ export class SurveyComponent implements OnInit {
...
-      const { solutionId, ...detailsOfProfile } = queryParam;
-      this.profileDetails = detailsOfProfile;
+      this.profileDetails = queryParam;
     });
-
     this.fetchSurveyDetails();
   }
+  openConfirmationDialog(confirmationParams: any): Promise<boolean> {
+    const dialogRef = this.dialog.open(DialogComponent, {
+      data: {
```
Comment (5Amogh) [inline kb-survey/src/app/survey/survey.component.ts]: instead of assigning each and every property we can just clone (deep copy instead of shallow copy) the `confirmationParams to data` will reduce a lot unnecessary code

File: kb-survey/src/app/survey/survey.component.ts [hunk 31]
```diff
@@ -101,30 +135,126 @@ export class SurveyComponent implements OnInit {
...
+              };
+
+              this.openConfirmationDialog(confirmationParams)
+            }
+          );
+      }
+    } else {
+      this.showSpinner = true;
+
+      this.baseApiService
```
Comment (5Amogh) [inline kb-survey/src/app/survey/survey.component.ts]: This is not resolved yet, you're rewriting entire logic for save and submit instead find a way to use the same logic for both cases

File: kb-survey/src/index.html [hunk 32]
```diff
@@ -7,6 +7,7 @@
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <link rel="icon" type="image/x-icon" href="favicon.ico">
   <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
+  <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@300;400;500&display=swap" rel="stylesheet">
```
Comment (5Amogh) [inline kb-survey/src/index.html]: why has this been added?

File: kb-survey/src/styles.scss (hunk 26 above)
Comment (5Amogh) [reply kb-survey/src/styles.scss]: still not  removed, pls check

**Review by 5Amogh — CHANGES_REQUESTED** [review]
Please re-check all comments again carefully and make the necessary changes

File: kb-survey/src/app/shared/components/dialog/dialog.component.html (hunk 28 above)
Comment (5Amogh) [reply kb-survey/src/app/shared/components/dialog/dialog.component.html]: cdkFocusInitial, this has not been resolved @syedameenabrar

File: kb-survey/src/index.html (hunk 32 above)
Comment (5Amogh) [reply kb-survey/src/index.html]: @syedameenabrar still not resolved

File: kb-survey/src/app/shared/components/dialog/dialog.component.scss [hunk 33]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+.mat-dialog-actions button {
+  padding: 7px 25px;
+  font-size: 15px;
+  border-radius: 17px;
+}
+
+
+
```
Comment (5Amogh) [inline kb-survey/src/app/shared/components/dialog/dialog.component.scss]: remove this long spaces between these

File: kb-survey/src/environments/environment.sample.ts
Comment (5Amogh) [inline kb-survey/src/environments/environment.sample.ts:1]: Why did you delete this file? @syedameenabrar this is needed.. remove only environment.ts that you'd have added for your local chanes

File: kb-survey/src/styles.scss (hunk 26 above)
Comment (5Amogh) [reply kb-survey/src/styles.scss]: No response on this yet and has not been changed

**Review by 5Amogh — CHANGES_REQUESTED** [review]

File: kb-survey/src/app/survey/survey.component.ts (hunk 20 above)
Comment (syedameenabrar) [reply kb-survey/src/app/survey/survey.component.ts:150]: it is done

File: kb-survey/src/styles.scss (hunk 26 above)
Comment (syedameenabrar) [reply kb-survey/src/styles.scss]: i have commented already, the unwanted code is removed

File: kb-survey/src/styles.scss
Comment (5Amogh) [inline kb-survey/src/styles.scss:1]: .mat-progress-spinner circle, .mat-spinner circle{
    stroke: var(--primary-color);
}
Add this code back, do not remove the old code

File: kb-survey/src/environments/environment.sample.ts
Comment (5Amogh) [inline kb-survey/src/environments/environment.sample.ts:1]: production: false,
  baseURL:'Insert your base url'
};
add this code back.. why are we making unnecessary changes? pls review your commits again

**Review by 5Amogh — CHANGES_REQUESTED** [review]

File: kb-survey/src/styles.scss
Comment (syedameenabrar) [reply kb-survey/src/styles.scss:1]: done

File: kb-survey/src/environments/environment.sample.ts
Comment (syedameenabrar) [reply kb-survey/src/environments/environment.sample.ts:1]: done

Comment (syedameenabrar) [reply kb-survey/src/environments/environment.sample.ts:1]: added

File: kb-survey/src/app/shared/components/dialog/dialog.component.html (hunk 28 above)
Comment (syedameenabrar) [reply kb-survey/src/app/shared/components/dialog/dialog.component.html]: i have removed
